dm/syncer/shardddl/optimist.go

```diff
@@ -156,6 +164,54 @@ func (o *Optimist) GetOperation(ctx context.Context, info optimism.Info, rev int
 	}
 }
 
+func (o *Optimist) GetRedirectOperation(ctx context.Context, info optimism.Info, rev int64) {
+	ctx2, cancel2 := context.WithCancel(ctx)
+
+	ch := make(chan optimism.Operation, 1)
+	errCh := make(chan error, 1)
+	targetTableID := utils.GenTableID(&filter.Table{Schema: info.DownSchema, Name: info.DownTable})
+	o.mu.Lock()
+	o.pendingRedirectCancelFunc[targetTableID] = cancel2
+	o.mu.Unlock()
+
+	go func() {
+		for {
+			op, rev2, err := optimism.GetOperation(o.cli, o.task, o.source, info.UpSchema, info.UpTable)
+			if err != nil {
+				o.logger.Warn("fail to get redirect operation", zap.Error(err))
+				time.Sleep(time.Second)
+				continue
+			}
+			// check whether operation is valid
+			if op.Task == o.task && rev2 > rev {
+				switch op.ConflictStage {
+				case optimism.ConflictResolved, optimism.ConflictNone:
+					o.saveRedirectOperation(targetTableID, &op)
+					return
+				}
+			}
+			ctx3, cancel3 := context.WithCancel(ctx2)
```

Comment:
do we need add a `defer cancel3` here?

Reply:
Every branch will employ `cancel3` now. There seems no need to run `cancel3` again.